bentoml/_internal/resource.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
+
+import os
+import re
+import math
+import typing as t
+import logging
+import functools
+from abc import ABCMeta
+from abc import abstractmethod
+
+import psutil
+
+from ..exceptions import BentoMLConfigException
+
+logger = logging.getLogger(__name__)
+
+_RESOURCE_REGISTRY: dict[str, t.Type[Resource[t.Any]]] = {}
+
+T = t.TypeVar("T")
+
+
+def get_resource(resources: dict[str, t.Any], resource_kind: str) -> t.Any:
+    if resource_kind not in _RESOURCE_REGISTRY:
+        raise BentoMLConfigException(f"Unknown resource kind '{resource_kind}'.")
+
+    resource: t.Type[Resource[t.Any]] = _RESOURCE_REGISTRY[resource_kind]
+
+    if resource_kind in resources:
+        if resources[resource_kind] == "system":
+            return resource.from_system()
+        else:
+            return resource.from_spec(resources[resource_kind])
+    else:
+        return None
+
+
+def system_resources() -> dict[str, t.Any]:
+    res: dict[str, t.Any] = {}
+    for resource_kind, resource in _RESOURCE_REGISTRY.items():
+        res[resource_kind] = resource.from_system()
+    return res
+
+
+class ResourceMeta(ABCMeta):
+    def __new__(  # pylint: disable=arguments-differ # we don't want to be generic
+        cls,
+        name: str,
+        bases: tuple[type],
+        attr_dict: dict[t.Any, t.Any],
+        *,
+        resource_id: str,
+    ):
+        res = super().__new__(cls, name, bases, attr_dict)
+        if resource_id != "":
+            # register the class with RESOURCE_REGISTRY
+            assert issubclass(res, Resource)
+            _RESOURCE_REGISTRY[resource_id] = res
+
+        return res  # pylint: disable=useless-super-delegation (???)
+
+
+class Resource(t.Generic[T], metaclass=ResourceMeta, resource_id=""):
```

Comment:
I feel like attrs class seems to be a nicer solution than having a metaclass. I guess the metaclass solution seems to be more magical (which i don't really have a strong opinion on) as this is more about how we want to maintain the code.

Reply:
This is to allow users to add custom resources; it functions effectively the same as the deploymentcontainer registry. An attrs class wouldn't work for that I think.

Reply:
I was thinking to use `attr.make_class` 🤔, but honestly it doesn't good typing.

Again, I don't have a strong opinion, but I'm curious to learn why meta classes will introduce limitation for custom runnable class @bojiang

Reply:
Correct me if im wrong, but having metaclass can introduce conflicts if user custom runnable class also implements/extends a metaclass, hence resolving metaclass in the case of multiple inheritance is not recommended.

 However, we should discourage using [multiple inheritance](https://www.youtube.com/watch?v=3MNVP9-hglc&ab_channel=NextDayVideo) anyways.

Reply:
How about use `__init_subclass__`?

```python
class Resource(t.Generic[T]):
    def __init_subclass__(cls, resource_id="", **kwargs):
        super().__init_subclass__(**kwargs)
        if resource_id == "":
            return
        # register the class with RESOURCE_REGISTRY
        _RESOURCE_REGISTRY[resource_id] = cls
```

Reply:
I believe we want to abstract away from initialization process. But from the breakout chat we decided to just use the old behaviour.

Reply:
That was for `Runnable`, not `Resource`; I've switched to `__init_subclass__`.